File: api-chi/cmd/services/movie_test.go
package services

import (
	"api-chi/cmd/models"

	"testing"

	"github.com/stretchr/testify/assert"
)

func Test_MovieService(t *testing.T) {
	id := ""
	service := MovieService{}

	t.Run("Count success", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Count database
		count, err := service.Count()
		assert.NoError(t, err)
		assert.Greater(t, count, 0)
	})

	t.Run("GetAll default success", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		search := ""
		limit := 10
		page := 1

		// Count database
		data, err := service.GetAll(&search, &limit, &page)
		assert.NoError(t, err)
		assert.IsType(t, data[0], models.Movie{})
		count := 0
		for _, item := range data {
			count += 1
			assert.NotEmpty(t, item.Id)
			assert.NotEmpty(t, item.Name)
		}
		assert.Equal(t, count, 10)
	})

	t.Run("GetAll limit < 10 will return 10 success", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		search := ""
		limit := 9
		page := 1

		// Count database
		data, err := service.GetAll(&search, &limit, &page)
		assert.NoError(t, err)
		assert.IsType(t, data[0], models.Movie{})
		count := 0
		for _, item := range data {
			count += 1
			assert.NotEmpty(t, item.Id)
			assert.NotEmpty(t, item.Name)
		}
		assert.Equal(t, count, 10)
	})

	t.Run("GetAll limit > 50 will return 50 success", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		search := ""
		limit := 51
		page := 1

		// Count database
		data, err := service.GetAll(&search, &limit, &page)
		assert.NoError(t, err)
		assert.IsType(t, data[0], models.Movie{})
		count := 0
		for _, item := range data {
			count += 1
			assert.NotEmpty(t, item.Id)
			assert.NotEmpty(t, item.Name)
		}
		assert.Equal(t, count, 50)
	})

	t.Run("GetAll page < 1 will return 50 success", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		search := ""
		limit := 10
		page := 0

		// Count database
		data, err := service.GetAll(&search, &limit, &page)
		assert.NoError(t, err)
		assert.IsType(t, data[0], models.Movie{})
		count := 0
		for _, item := range data {
			count += 1
			assert.NotEmpty(t, item.Id)
			assert.NotEmpty(t, item.Name)
		}
		assert.Equal(t, count, 10)
	})

	t.Run("GetAll with search is movie 1", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		search := "movie 1"
		limit := 20
		page := 1

		// Count database
		data, err := service.GetAll(&search, &limit, &page)
		assert.NoError(t, err)
		assert.IsType(t, data[0], models.Movie{})
		count := 0
		for _, item := range data {
			count += 1
			assert.NotEmpty(t, item.Id)
			assert.NotEmpty(t, item.Name)
		}
		assert.Less(t, count, 20)
	})

	t.Run("Create success with all attribute", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		input := models.Movie{
			Name:            "test movie",
			InitialDate:     "2024-10-12",
			ThumbnailImage:  "test.png",
			TrailerVideoUrl: "test-trailer",
			Description:     "movie is test movie",
		}

		// Create database
		value, err := service.Create(&input)
		assert.NoError(t, err)
		assert.NotEmpty(t, value)

		// Assign value to id
		id = value
	})

	t.Run("Update success with name", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		input := models.Movie{
			Id:   id,
			Name: "test movie is a good test movie",
		}

		// Update database
		value, err := service.Update(&input)
		assert.NoError(t, err)
		assert.NotEmpty(t, value)
		assert.Equal(t, value.Id, input.Id)
		assert.Equal(t, value.Name, input.Name)
	})

	t.Run("Update success with initial date", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		input := models.Movie{
			Id:          id,
			InitialDate: "2024-12-25",
		}

		// Update database
		value, err := service.Update(&input)
		assert.NoError(t, err)
		assert.NotEmpty(t, value)
		assert.Equal(t, value.Id, input.Id)
		assert.Equal(t, value.InitialDate, input.InitialDate)
	})

	t.Run("Update success with thumbnail image", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		input := models.Movie{
			Id:             id,
			ThumbnailImage: "movie-test.png",
		}

		// Update database
		value, err := service.Update(&input)
		assert.NoError(t, err)
		assert.NotEmpty(t, value)
		assert.Equal(t, value.Id, input.Id)
		assert.Equal(t, value.ThumbnailImage, input.ThumbnailImage)
	})

	t.Run("Update success with trailer video url", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		input := models.Movie{
			Id:              id,
			TrailerVideoUrl: "trailer-video-url-movie",
		}

		// Update database
		value, err := service.Update(&input)
		assert.NoError(t, err)
		assert.NotEmpty(t, value)
		assert.Equal(t, value.Id, input.Id)
		assert.Equal(t, value.TrailerVideoUrl, input.TrailerVideoUrl)
	})

	t.Run("Update success with description", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		input := models.Movie{
			Id:          id,
			Description: "this is description of test movie",
		}

		// Update database
		value, err := service.Update(&input)
		assert.NoError(t, err)
		assert.NotEmpty(t, value)
		assert.Equal(t, value.Id, input.Id)
		assert.Equal(t, value.Description, input.Description)
	})

	t.Run("Update success with all", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Declare input
		input := models.Movie{
			Id:              id,
			Name:            "this is a only test movie",
			InitialDate:     "2024-12-30",
			ThumbnailImage:  "thumbnail-for-best-test-movie.png",
			TrailerVideoUrl: "this-best-trailer-video-url-for-test-movie",
			Description:     "this move is description for test only movie",
		}

		// Update database
		value, err := service.Update(&input)
		assert.NoError(t, err)
		assert.NotEmpty(t, value)
		assert.Equal(t, value.Id, input.Id)
		assert.Equal(t, value.Name, input.Name)
		assert.Equal(t, value.InitialDate, input.InitialDate)
		assert.Equal(t, value.ThumbnailImage, input.ThumbnailImage)
		assert.Equal(t, value.TrailerVideoUrl, input.TrailerVideoUrl)
		assert.Equal(t, value.Description, input.Description)
	})

	t.Run("Remove success", func(t *testing.T) {
		// Connect database
		err := service.Open()
		defer service.Close()
		assert.NoError(t, err)

		// Remove database
		value, err := service.Remove(&id)
		assert.NoError(t, err)
		assert.NotEmpty(t, value)
	})
}

File: api-fastify/src/tests/services/movie.test.ts
import MovieService from "@/api/services/movie"
import { EnumMessage, ReturnMessage } from "@/types/message"
import { Movie } from "@/types/movie"

// Define test users
let currentId: string = ""

describe("MovieService Integration Tests", () => {
  let service: MovieService

  beforeAll(async () => {
    service = new MovieService()
  })

  describe("Count", () => {
    it("should return the count of movies", async () => {
      const result: ReturnMessage = await service.Count()
      expect(result.success).toBe(true)
      expect(result.message).toBe(EnumMessage.GET_SUCCESS)
      expect(typeof result.data).toBe("number")
      expect(result.data).toBeGreaterThan(0)
    })
  })

  describe("GetAll", () => {
    it("should return data of movies by default", async () => {
      const result: ReturnMessage = await service.GetAll("", 10, 1)
      expect(result.success).toBe(true)
      expect(result.message).toBe(EnumMessage.GET_SUCCESS)
      expect(Array.isArray(result.data)).toBe(true)
      let countData: number = 0
      result.data.forEach((movie: Movie) => {
        expect(typeof movie.id).toBe("string")
        expect(typeof movie.name).toBe("string")
        expect(typeof movie.initialDate).toBe("object");
        expect(typeof movie.thumbnailImage).toBe("string")
        expect(typeof movie.trailerVideoUrl).toBe("string")
        expect(typeof movie.description).toBe("string")
        expect(movie.id).not.toBe("")
        expect(movie.name).not.toBe("")
        expect(!isNaN(movie.initialDate.getTime())).toBe(true);
        countData += 1
      })
      expect(countData).toBe(10)
    })

    it("should return data of movie less than 20 values when searching", async () => {
      const result: ReturnMessage = await service.GetAll("category 1", 20, 1)
      expect(result.success).toBe(true)
      expect(result.message).toBe(EnumMessage.GET_SUCCESS)
      expect(Array.isArray(result.data)).toBe(true)
      let countData: number = 0
      result.data.forEach((movie: Movie) => {
        expect(typeof movie.id).toBe("string")
        expect(typeof movie.name).toBe("string")
        expect(typeof movie.initialDate).toBe("object");
        expect(typeof movie.thumbnailImage).toBe("string")
        expect(typeof movie.trailerVideoUrl).toBe("string")
        expect(typeof movie.description).toBe("string")
        expect(movie.id).not.toBe("")
        expect(movie.name).not.toBe("")
        expect(!isNaN(movie.initialDate.getTime())).toBe(true);
        countData += 1
      })
      expect(countData).toBeLessThanOrEqual(20)
    })

    it("should return data of movie with 10 values when limit = 9", async () => {
      const result: ReturnMessage = await service.GetAll("", 9, 1)
      expect(result.success).toBe(true)
      expect(result.message).toBe(EnumMessage.GET_SUCCESS)
      expect(Array.isArray(result.data)).toBe(true)
      let countData: number = 0
      result.data.forEach((movie: Movie) => {
        expect(typeof movie.id).toBe("string")
        expect(typeof movie.name).toBe("string")
        expect(typeof movie.initialDate).toBe("object");
        expect(typeof movie.thumbnailImage).toBe("string")
        expect(typeof movie.trailerVideoUrl).toBe("string")
        expect(typeof movie.description).toBe("string")
        expect(movie.id).not.toBe("")
        expect(movie.name).not.toBe("")
        expect(!isNaN(movie.initialDate.getTime())).toBe(true);
        countData += 1
      })
      expect(countData).toBe(10)
    })

    it("should return data of movie with 50 values when limit = 51", async () => {
      const result: ReturnMessage = await service.GetAll("", 51, 1)
      expect(result.success).toBe(true)
      expect(result.message).toBe(EnumMessage.GET_SUCCESS)
      expect(Array.isArray(result.data)).toBe(true)
      let countData: number = 0
      result.data.forEach((movie: Movie) => {
        expect(typeof movie.id).toBe("string")
        expect(typeof movie.name).toBe("string")
        expect(typeof movie.initialDate).toBe("object");
        expect(typeof movie.thumbnailImage).toBe("string")
        expect(typeof movie.trailerVideoUrl).toBe("string")
        expect(typeof movie.description).toBe("string")
        expect(movie.id).not.toBe("")
        expect(movie.name).not.toBe("")
        expect(!isNaN(movie.initialDate.getTime())).toBe(true);
        countData += 1
      })
      expect(countData).toBe(50)
    })

    it("should return data of movie when page = -1", async () => {
      const result: ReturnMessage = await service.GetAll("", 10, -1)
      expect(result.success).toBe(true)
      expect(result.message).toBe(EnumMessage.GET_SUCCESS)
      expect(Array.isArray(result.data)).toBe(true)
      let countData: number = 0
      result.data.forEach((movie: Movie) => {
        expect(typeof movie.id).toBe("string")
        expect(typeof movie.name).toBe("string")
        expect(typeof movie.initialDate).toBe("object");
        expect(typeof movie.thumbnailImage).toBe("string")
        expect(typeof movie.trailerVideoUrl).toBe("string")
        expect(typeof movie.description).toBe("string")
        expect(movie.id).not.toBe("")
        expect(movie.name).not.toBe("")
        expect(!isNaN(movie.initialDate.getTime())).toBe(true);
        countData += 1
      })
      expect(countData).toBe(10)
    })
  })

  describe("Create", () => {
    it("should create a movie successfully", async () => {
      const testCreated: Movie = {
        id: "",
        name: "test movie is 1",
        initialDate: new Date(),
        thumbnailImage: "image.png",
        trailerVideoUrl: "trailer-url",
        description: "description about test movie is 1"
      }
      const result: ReturnMessage = await service.Create(testCreated)
      expect(result.success).toBe(true)
      expect(result.message).toBe(EnumMessage.CREATE_SUCCESS)
      expect(result.data).not.toBeUndefined()
      expect(typeof result.data.id).toBe("string")
      expect(typeof result.data.name).toBe("string")
      expect(typeof result.data.initialDate).toBe("object");
      expect(typeof result.data.thumbnailImage).toBe("string")
      expect(typeof result.data.trailerVideoUrl).toBe("string")
      expect(typeof result.data.description).toBe("string")
      expect(result.data.id).not.toBe(undefined)
      expect(result.data.name).toBe(testCreated.name)
      expect(!isNaN(result.data.initialDate.getTime())).toBe(true);
      expect(result.data.thumbnailImage).toBe(testCreated.thumbnailImage)
      expect(result.data.trailerVideoUrl).toBe(testCreated.trailerVideoUrl)
      expect(result.data.description).toBe(testCreated.description)
      currentId = result.data.id
    })
  })

  describe("Update", () => {
    it("should update a movie successfully with id and name", async () => {
      const testUpdated: Movie = {
        id: currentId,
        name: "this is test movie 1",
        initialDate: new Date(),
        thumbnailImage: "test-movie-image.png",
        trailerVideoUrl: "test-movie-image-trailer",
        description: "test movie 1 description"
      }
      const result: ReturnMessage = await service.Update(testUpdated)
      expect(result.success).toBe(true)
      expect(result.message).toBe(EnumMessage.UPDATE_SUCCESS)
      expect(result.data).not.toBeUndefined()
      expect(typeof result.data.id).toBe("string")
      expect(typeof result.data.name).toBe("string")
      expect(typeof result.data.initialDate).toBe("object");
      expect(typeof result.data.thumbnailImage).toBe("string")
      expect(typeof result.data.trailerVideoUrl).toBe("string")
      expect(typeof result.data.description).toBe("string")
      expect(result.data.id).toBe(testUpdated.id)
      expect(result.data.name).toBe(testUpdated.name)
      expect(!isNaN(result.data.initialDate.getTime())).toBe(true);
      expect(result.data.thumbnailImage).toBe(testUpdated.thumbnailImage)
      expect(result.data.trailerVideoUrl).toBe(testUpdated.trailerVideoUrl)
      expect(result.data.description).toBe(testUpdated.description)
      currentId = result.data.id
    })
  })

  describe("Remove", () => {
    it("should remove a movie successfully", async () => {
      const resultBeforeCount: ReturnMessage = await service.Count()
      const result: ReturnMessage = await service.Remove(currentId)
      const resultAfterCount: ReturnMessage = await service.Count()
      expect(result.success).toBe(true)
      expect(result.message).toBe(EnumMessage.REMOVE_SUCCESS)
      expect(result.data).not.toBeUndefined()
      expect(resultBeforeCount.data).toBeGreaterThan(0)
      expect(resultAfterCount.data).toBeGreaterThan(0)
      expect(resultBeforeCount.data).toBeGreaterThan(resultAfterCount.data)
    })
  })
})
